theory/theory.md

```diff
@@ -0,0 +1,362 @@
+Source : https://www.geeksforgeeks.org/merge-sort/
+
+## 9. Generators - Generator functions allow you to declare a function that behaves like an iterator, i.e. it can be used in a for loop. What is the use case?
+
+
+Generator is useful when a machine’s memory is incapable of working a large dataset or when you have a complex function that needs to keep an internal state, but the function is too small to create a class. 
+
+#### Example 1 : Reading Large Files
+
+For example, Working with huge files, such as CSV files, or data streams is a frequent application for generators to avoid MemoryError. 
+
+    def csv_reader(file_name):
+        for row in open(file_name, "r"):
+            yield row
+    
+    # Rows are 64186394.
+
+The file is opened, iterated over, and a row is produced. The output from this code ought to be generated without any memory errors. So, in a sense, this program has converted csv reader() into a generator function. In this variation, a file is opened, each line is iterated through, and each row is yielded rather than returned.
+
+#### Example 2:  Generating an Infinite Sequence
+
+Infinite sequences can also be produced using a generator. Again, since memory is limited, yield over return will prevent a computer crash as you work through an infinite sequence.
+
+    def infinite_sequence():
+        num = 0
+        while True:
+            yield num
+            num += 1
+
+#### Example 3: Detecting Palindromes
+
+Generators can cycle through an unlimited series of numbers and determine whether the number created is the same forwards and backwards, making them excellent tools for palindrome detection.
+
+Please note that the code examples are taken from the source below.
+
+Source: https://realpython.com/introduction-to-python-generators/
+
+
+## 10. Decorators - A page for useful (or potentially abusive?) decorator ideas. What is the return type of the decorator?
+
+The function returns another function, usually applied as a function transformation using the @wrapper syntax. 
+
+#### Useful use of decorators include:
+
+Timing Functions : It will measure the time a function takes to run and print the time.
+Debugging Code:  It will print the arguments and its return value every time the function is called
+Slowing Down Code : It will sleep one second before it calls the decorated function. It may be useful when you want to rate-limit a function that continuously checks whether a resource, such as a web page, has changed
+User Logged In: Often used with a web framework, a web page can only be visible to logged in users
+
+#### Harmful decorators idea:
+
+ hide a backdoor : Hiding a backdoor and a module that collects users credentials and sends the data to a remote server
+
+Source: https://docs.python.org/3/glossary.html
+https://realpython.com/primer-on-python-decorators/#timing-functions
+https://www.bleepingcomputer.com/news/security/backdoored-python-library-caught-stealing-ssh-credentials/
+
+

```

Comment:
Q 10. 10 / 10
Impressive, Keiko! Also, thank you for providing some bad examples of bad case decorator implementation. ALso, very useful reference you have added.